src/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs

```diff
@@ -0,0 +1,412 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+// 
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.Identity.Client.Http;
+
+namespace Microsoft.Identity.Client.AppConfig
+{
+    /// <summary>
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public abstract class AbstractApplicationBuilder<T>
+        where T : AbstractApplicationBuilder<T>
+    {
+        internal AbstractApplicationBuilder(ApplicationConfiguration configuration)
+        {
+            Config = configuration;
+        }
+
+        internal ApplicationConfiguration Config { get; }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="httpClientFactory"></param>
+        /// <returns></returns>
+        public T WithHttpClientFactory(IMsalHttpClientFactory httpClientFactory)
+        {
+            Config.HttpClientFactory = httpClientFactory;
+            return (T)this;
+        }
+
+        internal T WithHttpManager(IHttpManager httpManager)
+        {
+            Config.HttpManager = httpManager;
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public T WithLoggingCallback(LogCallback loggingCallback)
+        {
+            if (Config.LoggingCallback != null)
+            {
+                throw new InvalidOperationException("LoggingCallback has already been set");
+            }
+
+            Config.LoggingCallback = loggingCallback;
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public T WithDebugLoggingCallback()
+        {
+            if (Config.LoggingCallback != null)
+            {
+                throw new InvalidOperationException("LoggingCallback has already been set");
+            }
+
+            Config.LoggingCallback = (level, message, pii) => { Debug.WriteLine($"{level}: {message}"); };
+            return (T)this;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public T WithTelemetryHandler(ITelemetryHandler telemetryHandler)
```

Comment:
#consistency:  logging is a delegate callback function.  this has an interface with one method on it. it would be great for these to feel the same for consistency.  should we have an interface in all cases (logging/telemetry) so that we can expand upon the interface if/as needed?  Or should these both be delegates?

Reply:
I also tend to lean towards this being a delegate publicly.  let's discuss in our meeting and we will decide.  thx.

---
In reply to: [245862415](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/761#discussion_r245862415) [](ancestors = 245862415)